### PR TITLE
Ensemble defect and subopt speed bug fix 

### PR DESCRIPTION
**Pull request by LunarFawn**
Broke Ensemble defect and mfe defect out from subopt and made their opwn function calls. Exposed those to Eterna and created tests in EternaDev.test.ts for thise. subopt takes 16secs and ED adn mfe ed take approve 40 sec now.

### CODE REVIEW

File: lib/NUPACK/emscripten/FullEnsemble.cpp [hunk 1]
```diff
@@ -180,20 +178,23 @@ FullAdvancedResult* FullEnsembleNoBindingSite(const std::string& seqString, int
...
     for ( int i = 0; i < suboptStructs.nStructs; i++ ) {
         oneDnaStruct currentStruct = suboptStructs.validStructs[i];
-
+        
         //get the secondary strucutre in dot paren notation 
-        std::string singlestructure = getDotParens(pseudoknotted, suboptStructs.seqlength, &currentStruct);
+        //old method
+        //std::string singlestructure = getDotParens(pseudoknotted, suboptStructs.seqlength, &currentStruct);
         
+        std::string singlestructure = getDotParens(pseudoknotted, suboptStructs.seqlength, &currentStruct);
```
Comment (luxaritas) [inline lib/NUPACK/emscripten/FullEnsemble.cpp:193]: Nothing changed here?

File: lib/CMakeLists.txt [hunk 2]
```diff
@@ -22,6 +22,7 @@ if (EXISTS "RNApuzzler/RNApuzzler")
 	add_subdirectory(RNApuzzler)
 endif()
 
+
```
Comment (luxaritas) [inline lib/CMakeLists.txt:25]: Nit: unnecessary whitespace change

File: lib/NUPACK/emscripten/FullEnsemble.cpp [hunk 3]
```diff
@@ -180,20 +178,23 @@ FullAdvancedResult* FullEnsembleNoBindingSite(const std::string& seqString, int
...
         if (i == 0)
         {
           //this is the first one so it is the mfe          
-          mfeStructure =  singlestructure;
+          mfeStructure =singlestructure;
```
Comment (luxaritas) [inline lib/NUPACK/emscripten/FullEnsemble.cpp:197]: Nit: unnecessary whitespace change

File: lib/NUPACK/emscripten/FullEnsemble.cpp [hunk 4]
```diff
@@ -209,38 +210,81 @@ FullAdvancedResult* FullEnsembleNoBindingSite(const std::string& seqString, int
...
+
+  auto autoSeqString = MakeCString(seqString);
+  char* string = autoSeqString.get();
+
+  auto autoMfeStructure = MakeCString(MfeStructure);
+  char* MfeStructureString = autoMfeStructure.get();
+
+  
+
+  //bool array for
```
Comment (luxaritas) [inline lib/NUPACK/emscripten/FullEnsemble.cpp:233]: Unfinished comment?

File: lib/NUPACK/emscripten/FullEnsemble.cpp [hunk 5]
```diff
@@ -209,38 +210,81 @@ FullAdvancedResult* FullEnsembleNoBindingSite(const std::string& seqString, int
...
+  auto autoSeqString = MakeCString(seqString);
+  char* string = autoSeqString.get();
+
+  auto autoMfeStructure = MakeCString(MfeStructure);
+  char* MfeStructureString = autoMfeStructure.get();
+
+  
+
+  //bool array for 
+  bool testList[2] = {false};
```
Comment (luxaritas) [inline lib/NUPACK/emscripten/FullEnsemble.cpp:234]: Why is this a bool array instead of just two variables (with names that clearly correspond to their purposes)?

File: lib/NUPACK/emscripten/Utils.cpp [hunk 6]
```diff
@@ -128,4 +128,153 @@ std::string getDotParens(bool pseudoknotted, const int seqlength, oneDnaStruct *
     {
        return singlestructure;
     }
+}
+
+std::string getDotParensNew(char* RNAseq, const int *thepairs, bool makeForEterna) {
```
Comment (luxaritas) [inline lib/NUPACK/emscripten/Utils.cpp:133]: Seems like this is not actually used anywhere?

File: src/eterna/folding/__tests__/NuPACKDev.test.ts [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+import SecStruct from 'eterna/rnatypes/SecStruct';
+//import {Oligo} from 'eterna/rnatypes/Oligo';
+
+function CreateFolder(type: any): Promise<Folder | null> {
+    return type.create();
+}
+
+/*
+
+test('NuPACK:suboptstructuresNoOligos', () => {
```
Comment (luxaritas) [inline src/eterna/folding/__tests__/NuPACKDev.test.ts:19]: Are we able to remove all the commented out tests? Otherwise, should they be (modified and) enabled?

File: lib/NUPACK/emscripten/FullEnsemble.cpp [hunk 8]
```diff
@@ -18,6 +18,9 @@ extern DBL_TYPE *pairPrPb;  //for pseudoknots
 extern double CUTOFF;
 extern int Multistranded;
 
+bool forEternaDigest = true;
+bool notforEternaDigest= false;
+
```
Comment (luxaritas) [inline lib/NUPACK/emscripten/FullEnsemble.cpp:23]: Unused?

File: src/eterna/folding/Folder.ts [hunk 9]
```diff
@@ -19,13 +20,16 @@ export interface FullEvalCache {
...
+    suboptStructures: string[];
+    suboptEnergyError: number[];
+    suboptFreeEnergy: number[];
+}
+
+export interface DefectResult {
     ensembleDefect: number;
     ensembleDefectNormalized: number;
     mfeDefect: number;
     mfeDefectNormalized: number;
```
Comment (luxaritas) [inline src/eterna/folding/Folder.ts:32]: Just to ask the question (as I don't thoroughly understand defect): Should these always be computed together? Or should we split these out even further and only calculate what the user asks for? Particularly the "normalized" versions seem trivial to compute and are straightforwardly derived from the "normal" defect - there's an argument to be had for having the caller (eg, the Eternascript using the values) compute these if they're needed.

Also as a note, looking at the defect code, I'm wondering if more of this should just be done in TS to begin with. In general, I think we want to only do the minimum amount necessary in C++ (where performance matters or we actually have to call C/C++ functions), plus it looks like you're re-computing things like MFE and energy which we may already have cached in TS.

File: lib/NUPACK/emscripten/Utils.cpp (hunk 6 above)
Comment (LunarFawn) [reply lib/NUPACK/emscripten/Utils.cpp:133]: I can take that out. I had that in there trying to see if there was a speed difference between what we are currently doing and the method I had pulled frm nupack. did not see a difference really maybe like a fraction of a second in our current methods favor so I did not put any more effort into it and forgot I had that there. I will remove next commit.

File: src/eterna/folding/__tests__/NuPACKDev.test.ts (hunk 7 above)
Comment (LunarFawn) [reply src/eterna/folding/__tests__/NuPACKDev.test.ts:19]: I left them there for future troubleshooting help but I can pull those out np as they are covered in folder.test.ts i think

File: lib/NUPACK/emscripten/FullEnsemble.cpp (hunk 8 above)
Comment (LunarFawn) [reply lib/NUPACK/emscripten/FullEnsemble.cpp:23]: This is used for the newdotparens stuff and will be taken out when I pull that function out.

File: src/eterna/folding/Folder.ts (hunk 9 above)
Comment (LunarFawn) [reply src/eterna/folding/Folder.ts:32]: the call for FullEnsembleDefect inFullEnsemble.cpp has a variable called "mode" that is a integer that tells the function to find normal=1, mfe=2, or both=3. The external interface DefectResult  right now is coded up to return a 0 or -1 or a particular defect value if you did not ask for it / it was not computed. 

good point. there are a couple calls for structures and MFE. If we want to shift to more being done in TS and less in C++ then I can just make sure the nupack calls we need are exposed and then do those calls or grab from cache if available for that call in the future. You also make a good point about seperating out the ensemble defect portions out.

File: lib/NUPACK/emscripten/FullEnsemble.cpp (hunk 5 above)
Comment (LunarFawn) [reply lib/NUPACK/emscripten/FullEnsemble.cpp:234]: I think I was playing around with an idea and when it worked I forgot to change the names although I think I meant testList as in a list of thigs to test... but I see the confusion and will get changed

File: lib/NUPACK/emscripten/FullEnsemble.cpp (hunk 4 above)
Comment (LunarFawn) [reply lib/NUPACK/emscripten/FullEnsemble.cpp:233]: this was to explain what was going on with the mode stuff to explain why testList is testList

File: lib/CMakeLists.txt (hunk 2 above)
Comment (LunarFawn) [reply lib/CMakeLists.txt:25]: the new version that checks if exists is implemented now so this is no longer relevant.

File: lib/NUPACK/emscripten/FullEnsemble.cpp (hunk 1 above)
Comment (LunarFawn) [reply lib/NUPACK/emscripten/FullEnsemble.cpp:193]: this is left over from me tryin gout the other dot parens process to see if it gave a time boost that was worth putting time into understanding for implementation. Those comments are gone now.

File: lib/CMakeLists.txt (hunk 2 above)
Comment (LunarFawn) [reply lib/CMakeLists.txt:25]: took out the extra line

File: lib/NUPACK/emscripten/FullEnsemble.cpp (hunk 3 above)
Comment (LunarFawn) [reply lib/NUPACK/emscripten/FullEnsemble.cpp:197]: changed back as it is better with the space anyways

File: lib/NUPACK/emscripten/FullEnsemble.cpp (hunk 5 above)
Comment (LunarFawn) [reply lib/NUPACK/emscripten/FullEnsemble.cpp:234]: i made this two variables and got rid of extra code

File: lib/NUPACK/emscripten/Utils.cpp (hunk 6 above)
Comment (LunarFawn) [reply lib/NUPACK/emscripten/Utils.cpp:133]: i took this out and all references

File: src/eterna/folding/__tests__/NuPACKDev.test.ts (hunk 7 above)
Comment (LunarFawn) [reply src/eterna/folding/__tests__/NuPACKDev.test.ts:19]: i took them out

File: lib/NUPACK/emscripten/FullEnsemble.cpp (hunk 8 above)
Comment (LunarFawn) [reply lib/NUPACK/emscripten/FullEnsemble.cpp:23]: removed

File: src/eterna/folding/Folder.ts (hunk 9 above)
Comment (LunarFawn) [reply src/eterna/folding/Folder.ts:32]: I see some places were I can easily do some stuff in TS and it is harder in in others.What I intend to do at a minimum is to write it so that in TS it checks for a cached MFE struct and if available it will pass that to this function. This implementation then naturally leads to a separate ensembleDefect result and a mfe defect result in TS per @luxaritas previous comment about  further breaking up ED and MFE_ED results. As such I will break out ED and MFE_ED into separate calls and functions
